targets/TARGET_STMBLUE/DTM_libraries/BlueNrgHCIDriver.cpp

```diff
@@ -0,0 +1,390 @@
+                // handle extended command
+                if (hciCoreCb.extResetSeq)
+                {
+                    /* send next extended command in sequence */
+                    (*hciCoreCb.extResetSeq)(pMsg, opcode);
+                }
+                break;
+
+            case HCI_OPCODE_LE_RAND:
+                /* check if need to send second rand command */
+                if (randCnt < (HCI_RESET_RAND_CNT-1))
+                {
+                    randCnt++;
+                    HciLeRandCmd();
+                }
+                else
+                {
+                    /* last command in sequence; set resetting state and call callback */
+                    signal_reset_sequence_done();
+                }
+                break;
+
+            default:
+                break;
+        }
+    }
+
+    virtual ble::vendor::cordio::buf_pool_desc_t get_buffer_pool_description() {
+    	//return ble::vendor::cordio::CordioHCIDriver::get_default_buffer_pool_description();
+
+    	//minimal configuration for HRM example
+        uint8_t buffer[1430];
+        static const wsfBufPoolDesc_t pool_desc[] = {
+            {  16, 14 },
+            {  32, 10 },
+            {  64, 4 },
+            { 128, 2 },
+            { 272, 1 }
+        };
+
+        return ble::vendor::cordio::buf_pool_desc_t(buffer, pool_desc);
+    }
+
+private:
+
+    void hciCoreReadMaxDataLen(void)
+    {
+        /* if LE Data Packet Length Extensions is supported by Controller and included */
+        if ((hciCoreCb.leSupFeat & HCI_LE_SUP_FEAT_DATA_LEN_EXT) &&
+            (hciLeSupFeatCfg & HCI_LE_SUP_FEAT_DATA_LEN_EXT))
+        {
+            /* send next command in sequence */
+            HciLeReadMaxDataLen();
+        }
+        else
+        {
+            /* send next command in sequence */
+            HciLeRandCmd();
+        }
+    }
+
+    void hciCoreReadResolvingListSize(void)
+    {
+        /* if LL Privacy is supported by Controller and included */
+        if ((hciCoreCb.leSupFeat & HCI_LE_SUP_FEAT_PRIVACY) &&
+            (hciLeSupFeatCfg & HCI_LE_SUP_FEAT_PRIVACY))
+        {
+            /* send next command in sequence */
+            HciLeReadResolvingListSize();
+        }
+        else
+        {
+            hciCoreCb.resListSize = 0;
+
+            /* send next command in sequence */
+            hciCoreReadMaxDataLen();
+        }
+    }
+};
+
+/**
+ *  Virtual Transport driver, used to exchange packet between host and controller.
+ */
+class TransportDriver : public cordio::CordioHCITransportDriver {
+public:
+
+    TransportDriver() {}
+
+    virtual ~TransportDriver(){ }
+
+    /**
+     * @see CordioHCITransportDriver::initialize
+     */
+    virtual void initialize() {
+    	/* Stack Initialization */
+    	DTM_StackInit();
+    	////stacktick queueing, the dispatch loop will be initialized after the initialization
+    	////inside main
+    	_eventQueue = mbed_event_queue();
```

Comment:
It feels inefficient to pass the default event queue here: If the application doesn't use it like it is the case in *all* ble examples then you pay for two event queues. 
I need some time to think of a way of hooking an undefined recurring event in the Cordio port.

Reply:
Hi @pan-, rather than in the Cordio it could be better to hook the stack tick inside the dispatch loop of the event queue. I've started studying sleep and deep sleep capabilities of BlueNRG2, discovering that StackTick is also mandatory to get permission from controller to enter sleep mode.

Reply:
From my understanding BTLE_StackTick has to be called at every system-wakeup to keep up the 500ppm accuracy. There is an integration point in the sleep function. 

Going back to attaching it to the default event queue; it is wrong: `mbed_event_queue` creates a separate thread to that will dispatch events. This is not good resources wise; users must be able to execute their application on a single thread. 

BLE API itself doesn't assume the existence of an event queue of any sort: The application must register a callback that will signal the application that BLE must process. That signal can be tied to any type of event queue, a loop, ... It is up to the application. 

I'd like to extend the HCI driver as such: 
- Expose a function that can be called by the driver to signal their is work to do. 
- Expose a virtual function that process driver events and that can be overridden by the driver. It would be called whenever the bluetooth subsystem is processed.

```c++
// In CordioHCIDriver declaration

// invoke the private virtual function do_process_events();
void process_events(); 

// to call by the driver whenever there is an event. 
void signal_events_to_process();

private: 
// by default event processing does nothing 
virtual void do_process_events() {} 
```

In the ST driver this can be use as follow: 

```c++

virtual void initialise() { 
	/* Stack Initialization */
     	DTM_StackInit();

        // setup local Ticker for BTLE_StackTick
        _ticker.attach(mbed::callback(this, signal_events_to_process), 10000 /*us*/);
}

virtual void do_process_events() { 
    BTLE_StackTick();
}
```

@donatieng @paul-szczepanek-arm What do you think ?

Reply:
@pan- BTLE_StackTick performs the whole link layer and radio processing, it shall be called repeatedly to ensure the correct BLE controller operation (among the others to keep the 500 ppm accuracy, but it is not sufficient to call this signal only once after wakeup). Your proposed approach could be, in my opinion, ideal to ensure the correct processing and at the same time hide this mechanism to the application, with the actual BlueNRG2 controller library implementation.

Concerning the event queue, I had to modify the HRM example to use also there the pointed by the return pointer of `mbed_event_queue`, by changing with

`static EventQueue* event_queue = mbed_event_queue();`